src/goLiveErrors.ts

```diff
@@ -0,0 +1,100 @@
+'use strict';
+
+import vscode = require('vscode');
+import { byteOffsetAt, getBinPath } from './util';
+import cp = require('child_process');
+import path = require('path');
+import { promptForMissingTool } from './goInstallTools';
+import { diagnosticCollection } from './goMain';
+
+// Interface for settings configuration for adding and removing tags
+interface GoLiveErrorsConfig {
+	delay: number;
+	enabled: boolean;
+}
+
+let runner;
+
+// parseLiveFile runs the gotype command in live mode to check for any syntactic or
+// semantic errors and reports them immediately
+export function parseLiveFile(e: vscode.TextDocumentChangeEvent) {
+	if (e.document.isUntitled) {
+		return;
+	}
+	if (e.document.languageId !== 'go') {
+		return;
+	}
+
+	let config = <GoLiveErrorsConfig>vscode.workspace.getConfiguration('go')['liveErrors'];
+	if (config == null || !config.enabled) {
+		return;
+	}
+
+	if (runner != null) {
+		clearTimeout(runner);
+	}
+	runner = setTimeout(function(){
+		processFile(e);
+		runner = null;
+	}, config.delay);
+}
+
+export function goLiveErrorsEnabled() {
+	return <GoLiveErrorsConfig>vscode.workspace.getConfiguration('go')['liveErrors'].enabled;
+}
+
+// processFile does the actual work once the timeout has fired
+function processFile(e: vscode.TextDocumentChangeEvent) {
+	let uri = e.document.uri;
+	let gotypeLive = getBinPath('gotype-live');
+	let fileContents = e.document.getText();
+	let fileName = e.document.fileName;
+	let args = ['-e', '-a', '-lf=' + fileName, path.dirname(fileName)];
+	let p = cp.execFile(gotypeLive, args, (err, stdout, stderr) => {
+		if (err && (<any>err).code === 'ENOENT') {
+			promptForMissingTool('gotype-live');
+			return;
+		}
+
+		// we want to take the error path here because the command we are calling
+		// returns a non-zero exit status if the checks fail
+		let newDiagnostics = [];
+		if (!newDiagnostics) {
+			newDiagnostics = [];
+		}
+
+		// grab a copy of the existing diagnostics that are being reported for the
+		// current file
+		let oldDiagnostics = diagnosticCollection.get(uri);
+
+		// delete the existing diagnostics for the current file
+		//
+		// error-level diagnostics will be reported by this process, so we want to
+		// clear out the existing errors to avoid getting duplicates
+		diagnosticCollection.delete(uri);
```

Comment:
I am not too picky about this one though. Perf wise, it shouldn't be a problem because you are only touching diagnostics for a single file. But it just feels wrong (gut wise) to be copying array items everytime liveError is run